src/main/java/com/expertsoft/tasks/OrderStats.java

```diff
@@ -1,120 +1,148 @@
      * If customer stream is empty, Optional.empty should be returned.
-     *
+     * <p>
      * Example: For the stream, containing
-     *      Customer#1 -> USA
-     *      Customer#2 -> France
-     *      Customer#3 -> Japan
-     *      Customer#4 -> USA
-     *      Customer#5 -> Japan
-     *
-     *      "USA" should be returned.
+     * Customer#1 -> USA
+     * Customer#2 -> France
+     * Customer#3 -> Japan
+     * Customer#4 -> USA
+     * Customer#5 -> Japan
+     * <p>
+     * "USA" should be returned.
      *
      * @param customers stream of customers
      * @return java.util.Optional containing the name of the most popular country
      */
     static Optional<String> mostPopularCountry(final Stream<Customer> customers) {
-        return null;
+        String result;
+        try {
+            result = customers
+                    .collect(Collectors.groupingBy(customer -> customer.getAddress().getCountry(), Collectors.counting()))
+                    .entrySet()
+                    .stream()
+                    .max(Map.Entry.comparingByValue())
+                    .get()
+                    .getKey();
+        } catch (NoSuchElementException e) {
+            return Optional.empty();
+        }
+        return Optional.of(result);
     }
 
     /**
      * Task 6 (⚫⚫⚫⚫⚫)
-     *
+     * <p>
      * Given a stream of customers, return the average product price for the provided credit card number.
-     *
+     * <p>
      * Info: If order contains the following order items:
-     *  [
-     *      Product1(price = 100$, quantity = 2),
-     *      Product2(price = 160$, quantity = 1)
-     *  ]
+     * [
+     * Product1(price = 100$, quantity = 2),
+     * Product2(price = 160$, quantity = 1)
+     * ]
      * then the average product price for this order will be 120$ ((100 * 2 + 160 * 1) / 3)
-     *
+     * <p>
      * Hint: Since product prices are represented as BigDecimal objects, you are provided with the collector implementation
-     *       to compute the average value of BigDecimal stream.
+     * to compute the average value of BigDecimal stream.
      *
-     * @param customers stream of customers
+     * @param customers  stream of customers
      * @param cardNumber card number to check
      * @return average price of the product, ordered with the provided card
      */
     static BigDecimal averageProductPriceForCreditCard(final Stream<Customer> customers, final String cardNumber) {
         final AveragingBigDecimalCollector collector = new AveragingBigDecimalCollector();
-        return null;
+        Stream<BigDecimal> prices = customers.flatMap(customer -> customer.getOrders().stream())
+                .filter(order -> order.getPaymentInfo().getCardNumber().equals(cardNumber))
+                .flatMap(order -> order.getOrderItems().stream())
+                .flatMap(orderItem -> Collections.nCopies(orderItem.getQuantity(), orderItem.getProduct()).stream())
```

Comment:
not sure why do we need list of copies with size = quantity

Reply:
due to this comment in requirements:
--------------------------------------------------------------
     * [
     * Product1(price = 100$, quantity = 2),
     * Product2(price = 160$, quantity = 1)
     * ]
     * then the average product price for this order will be 120$ ((100 * 2 + 160 * 1) / 3)
-------------------------------------------------------------
we need to multiply items on quantity